src/lava/magma/core/model/py/connection.py

```diff
@@ -0,0 +1,1702 @@
+                               self._learning_rule.y2_impulse,
+                               self._learning_rule.y3_impulse])
+        self._y_impulses_int, self._y_impulses_frac = \
+            self._decompose_impulses(y_impulses)
+        self._y_taus = np.array(
+            [
+                self._learning_rule.y1_tau,
+                self._learning_rule.y2_tau,
+                self._learning_rule.y3_tau,
+            ]
+        )
+
+    def _build_active_traces_per_dependency(self) -> None:
+        """Build and store boolean numpy arrays specifying which x and y
+        traces are active, per dependency.
+
+        First dimension:
+        index 0 -> x0 dependency
+        index 1 -> y0 dependency
+        index 2 -> u dependency
+
+        Second dimension (for x_traces):
+        index 0 -> x1 trace
+        index 1 -> x2 trace
+
+        Second dimension (for y_traces):
+        index 0 -> y1 trace
+        index 1 -> y2 trace
+        index 2 -> y3 trace
+        """
+        # Shape : (3, 5)
+        active_traces_per_dependency = np.zeros(
+            (
+                len(str_symbols.DEPENDENCIES),
+                len(str_symbols.PRE_TRACES) + len(str_symbols.POST_TRACES),
+            ),
+            dtype=bool,
+        )
+        for (
+                dependency,
+                traces,
+        ) in self._learning_rule.active_traces_per_dependency.items():
+            if dependency == str_symbols.X0:
+                dependency_idx = 0
+            elif dependency == str_symbols.Y0:
+                dependency_idx = 1
+            elif dependency == str_symbols.U:
+                dependency_idx = 2
+            else:
+                raise ValueError("Unknown Dependency in ProcessModel.")
+
+            for trace in traces:
+                if trace == str_symbols.X1:
+                    trace_idx = 0
+                elif trace == str_symbols.X2:
+                    trace_idx = 1
+                elif trace == str_symbols.Y1:
+                    trace_idx = 2
+                elif trace == str_symbols.Y2:
+                    trace_idx = 3
+                elif trace == str_symbols.Y3:
+                    trace_idx = 4
+                else:
+                    raise ValueError("Unknown Trace in ProcessModel")
+
+                active_traces_per_dependency[dependency_idx, trace_idx] = True
+
+        # Shape : (3, 2)
+        self._active_x_traces_per_dependency = active_traces_per_dependency[
+                                               :, :2
+                                               ]
+
+        # Shape : (3, 3)
+        self._active_y_traces_per_dependency = active_traces_per_dependency[
+                                               :, 2:
+                                               ]
+
+    def _build_active_traces(self) -> None:
+        """Build and store boolean numpy arrays specifying which x and y
+        traces are active."""
+        # Shape : (2, )
```

Comment:
What does this shape refer to? Is it the shape of `self._active_x_traces`? Why is it important to know the shape here?

Reply:
Yes, I think so. Since the shape of self._active_x_traces_per_dependency is [3, 2] - I believe a logical_or of each of its elements (which is [2, ]) would return an array of size [2, ]. So shape of self._active_x_traces is [2,]

Reply:
We restructured the code for more readability.